### ISSUE TITLE
window.print() only works once

### ISSUE THREAD

**Issue by michaelsmitha** [electron/electron#14705]
* Output of `node_modules/.bin/electron --version`: _v3.0.0-beta.13_
* Operating System (Platform and Version): _Windows 10_
* Output of `node_modules/.bin/electron --version` on last known working Electron version (if applicable): _?_

**Expected Behavior**
window.print() should print the window on each subsequent call

**Actual behavior**
window.print() will work once, but fail from there on. This looks like a regression of https://github.com/electron/electron/pull/1532? 
The device selection always comes up successfully. I've tried printing both to "Microsoft Print to PDF", which only gives you the "save location" dialog once, and my local printer, which will print once, then does nothing.

**To Reproduce**
_Create a barebones electron app that can call `window.print()` as desired. Repository included._

clone: https://github.com/michaelsmitha/print-test

* Run `npm install`
* Run `.\node_modules\electron\dist\electron.exe .`
* Clicking on the document will trigger a window.print()

Attempting to print the window after the first time fails.

**Comment by jrobeson** [comment]
what version of electron?

**Comment by maddhruv** [comment]
@michaelsmitha works fine for me every time
electron version - v2.0.10

**Comment by michaelsmitha** [comment]
Right, it doesn't in v3.0.0-beta.13

**Comment by kevinsalerno** [comment]
Same here. Printer works fine on a fresh restart, very inconsistent with following calls.

**Windows 10**
"electron": "^3.0.0"

**Comment by ryansolida** [comment]
Anyone know of any workarounds for the moment on this bug?

**Comment by damianaiamad** [comment]
I'm seeing this bug in 3.0.4, Windows only.

**Comment by daikiuk** [comment]
Total showstopper for us, must downgrade to 2.0

**Comment by damianaiamad** [comment]
> Anyone know of any workarounds for the moment on this bug?

Closest I can suggest is to use webContents.printToPDF(), save to temp file, then shell.openExternal(). Then user has to print the PDF. Not a great workaround but kinda sorta gets you there...

**Comment by damianaiamad** [comment]
Retested with Electron 4.0.0-beta.4 / Windows 10: app now crashes after clicking Print button on print dialog.

**Comment by damianaiamad** [comment]
Retested with Electron 4.0.0-beta.4 / Windows 10: app now crashes after clicking Print button on print dialog.

**Comment by damianaiamad** [comment]
Retested with Electron 4.0.0-beta.4 / Windows 10: app now crashes after clicking Print button on print dialog.

**Comment by damianaiamad** [comment]
Retested with Electron 4.0.0-beta.4 / Windows 10: app now crashes after clicking Print button on print dialog.

**Comment by damianaiamad** [comment]
Retested with Electron 4.0.0-beta.4 / Windows 10: app now crashes after clicking Print button on print dialog.

**Comment by BeatriceThalo** [comment]
I'm on electron 1.8.8, here is a bit of off-topic rambling, which hopefully would spark some ideas as to what the cause may or may not be.

Before learning about ipc, I had some main-side print code:
client.ts
```
document.body.addEventListener('my-doc-rendered', (e: CustomEvent) => {
  window.open('', '_self'); // Call custom handler to print outside of angular, synchronous
  window.history.back();
}, {once: true});
```
electron.js
```
brWin.webContents.on('new-window', (event, urlParam, frameName) => {
  if (frameName === '_self') {
    event.preventDefault();
    win.webContents.print(); // synchronous (?)
  }
});
```
The actual behavior was happily blocking and synchronous for either a Cancel or Print choice in the print dialog. Windows and Ubuntu.

Then I learned about ipc and refactored to use renderer-side code:
client.ts
```
document.body.addEventListener('my-doc-rendered', (e: CustomEvent) => {
  window.print(); // asynchronous, w3schools says no arguments or return value
  window.close();
}, {once: true});
```
The observed behavior on Windows 10 is that the printDialog opens, then (likely before or) immediately after the user choice is made, the close method is executed, then (I imagine the blackbox as) printToPrinter or printToFile method is called, and creates a print job as seen in the printer's queue momentarily or does save a pdf file, but opening the pdf says something along the lines that it is empty or wrong, and the physical printer never spits out any paper.

In my attempts to refactor with the correct synchronicity, I went back to the electron way still on the renderer-side,
client.ts
```
document.body.addEventListener('my-doc-rendered', (e: CustomEvent) => {
  window.require('electron').remote.getCurrentWebContents().print(); // asynchronous
  window.close();
}, {once: true});
```
Back on Ubuntu 18.04, the browser window closes before the print dialog is ever seen. Not sure if my event chain made it seem synchronous before, or the non-remote method is different than the remote one.

Utilizing the callback,
client.ts
```
document.body.addEventListener('my-doc-rendered', (e: CustomEvent) => {
  window.require('electron').remote.getCurrentWebContents().print({}, (success: boolean) => { // my linter requires types and arrow syntax, idk if it matters
    console.log('success?', success) // no semicolon, idk if it matters
    window.close();
  });
  console.log('after print call line'); // shows in console behind open print dialog
}, {once: true});
```
The print dialog opened, I chose Cancel or Print to file maybe, the dialog closed, the window stayed open, and my entire machine froze, needing to be forcibly shut down. I was also running Slack in a Firefox tab at the time.

With no Slack tab open, this works as expected when the Print choice is made, but the callback does not execute if the Cancel choice is made. So if that is intended, I recommend changing the webContents documentation to refer to it as a successCallback instead of a callback with a success/error boolean argument.

**Comment by lubo08** [comment]
electron 5.0.5 again this bug is there in 3.1.8 wasn't

**Comment by paulbennet** [comment]
Also happening in electron 5.0.7 .. Fyi

**Comment by jainprashul** [comment]
Happening in electron 5.0.8

**Comment by mustafahafidi** [comment]
Happening in electron 6.0.0

**Comment by aneesijaz** [comment]
It's also in electron 5.0.6 👎

**Comment by paulbennet** [comment]
I guess we should open a new issue for this instead of commenting on an year old conversation ?!

**Comment by purocean** [comment]
6.0.0 too

**Comment by Jaifroid** [comment]
After issuing window.print() inside an electron app, the app crashes after user has chosen "print" in the dialog box. It sometimes also crashes after user chooses Cancel. Windows 10.

**Comment by vkiranmaniya** [comment]
Application is crashing after print or save as pdf in v6.0.1 too. have you found any solution?

**Comment by mustafahafidi** [comment]
Could we just open a new issue instead??

**Comment by patrickrobbins** [comment]
I'm still experiencing this issue in win 10 Electron 5.x

**Comment by courageDeveloper** [comment]
Was previously using electron version 6 and my processes crashed immediately after printing. I had to downgrade to electron 2.0.2.  One thing i noticed is that in electron 2.0.2, i had to !-- Insert this line above script imports  -->
<script>if (typeof module === 'object') {window.module = module; module = undefined;}</script>
and also 

<script>if (window.module) module = window.module;</script>
for most cases involving using Jquery scripts or Javascript. I thought about what if in Version 6 adding those scripts above could probably stop the process from crashing. You can try it out before downgrading, as for me i have a projects to deliver so i will just stick with my downgraded version as it works fine, at least for now and hopefully through deployment.

**Comment by TristonJ** [comment]
This is a pretty old issue, but I am running into it on electron `7.0.0-beta.7` (Linux) and I am unable to execute `window.print()` more than once.

I have a workaround that appears to function in my case. You can spawn a new hidden `BrowserWindow`, load the HTML / CSS to be printed, and execute `window.print(); setTimeout(() => window.close());` on that window. For example: 
```js
const printRawHtml = html => {
  const window = new BrowserWindow({ show: false });
  window.loadURL(`data:text/html;charset=UTF-8,${
    encodeURIComponent(`<html><body>${html}</body></html>`)
  }`);
  window.webContents.on('did-finish-load', () => {
    window.webContents.executeJavaScript(
      'window.print(); setTimeout(() => window.close());',
    );
  });
};
```
This of course only works if you are easily able to retrieve the necessary css/html from the page you would like to be printed. This was only tested on Ubuntu 19.04.

**Comment by DoYourUX** [comment]
This is still an issue in 7.0.0

**Comment by barbalex** [comment]
Yep, I can reproduce this in 7.0.0 on Windows

**Comment by avivsalman** [comment]
Any updates for this bug? I am in production and it's still in 7.0.0 on windows...

**Comment by InfiniteLukeOne** [comment]
Having this issue in electron v7.1.1. Printer selection window is displayed after the first try, but printing only works the first time.

Steps to reproduce like in original Post with a npm install electron@latest before starting.

**Comment by MZiganow** [comment]
Still an issue in electron v7.1.2 on Windows 10. Same behavior as described by @InfiniteLukeOne

**Comment by InfiniteLukeOne** [comment]
> 
> 
> Still an issue in electron v7.1.2 on Windows 10. Same behavior as described by @InfiniteLukeOne

Take a look at #21195. There's a workaround in the comments which worked for me.

**Comment by dvuvne** [comment]
I am facing exactly the same issue. The only workaround that worked for me was downgrading to electron version 3.

**Comment by peacegroup** [comment]
> Same here. Yes this is a bug in v3 on Windows. Printer works fine on a fresh restart, very inconsistent with following calls.
> 
> **Windows 10**
> "electron": "^3.0.0"

> Ubuntu 18 experiences the same issue

**Comment by Suroor-Ahmmad** [comment]
Is this issue resolved? I'm facing same issue in Electron 10.1.0

**Comment by karlitos** [comment]
I face the same issue in Electron 9.1.2. on MacOs with the `printToPDF` method

**Comment by patrickrobbins** [comment]
I got around this by capturing the screen to pdf, and reprinting that. It
is a small use case but worked for my work flow

**Comment by Jaifroid** [comment]
It was solved for me in Electron 9.0.5 when I last implemented this in an Electron version of the WikiMed app. There I can print as many times as I like without a crash (just tested the same app again). I hope they haven't re-introduced the bug since 9.0.5.

**Comment by karlitos** [comment]
> It was solved for me in Electron 9.0.5 when I last implemented this in an Electron version of the WikiMed app. There I can print as many times as I like without a crash (just tested the same app again). I hope they haven't re-introduced the bug since 9.0.5.

I just tried with electron 9.0.5. and I was able to print multiple times to PDF without any problems.

**Comment by Suroor-Ahmmad** [comment]
Can someone check on Electron 10.1.0, I'm facing similar kind of issue, opened separate issue and it's tagged as bug #25327

**Comment by mad1009** [comment]
A work around for previous versions, based On @TristonJ answer

```javascript
  // Create a new window to handle printing
  const printWin = new BrowserWindow({ show: false });
  // Load current window url
  printWin.loadURL(BrowserWindow.getFocusedWindow().webContents.getURL());
  // Make sure the content fully loaded
  printWin.webContents.on('did-finish-load', () => {
      printWin.on("close",()=>{
        console.log("print window closed")
      })
      // Handle printing logic and notify the mainWindow
      printWin.webContents.print(printingOptions,(success, err) => {
          if(success){
            mainWindow.webContents.send("print:success")
            printWin.close()
          }else{
            mainWindow.webContents.send("print:failed",{error:err})
            printWin.close()
          }
          console.log('err :>> ', err);
        });
  });
```
When printing on the same window, ```mainWindow.print()``` will work only one time, and the callback will always return a success value.
To avoid this bug, before printing always create a hidden window, load the current url, and print.
Note that in the callback function the print window always notify the main window about the print status and close the print window to avoid memory leaks.
Note: When printing an HTML page make sure to use ``` @media print{...} ``` and other css tools to controle your printing view
